modules/engine/src/shader-inputs.ts

```diff
@@ -61,17 +61,20 @@ export class ShaderInputs<
    * @param modules
    */
   // @ts-expect-error Fix typings
-  constructor(modules: {[P in keyof ShaderPropsT]: ShaderModuleInputs<ShaderPropsT[P]>}) {
-    // TODO - get all dependencies from modules
-    const allModules = _resolveModules(Object.values(modules));
-    log.log(
-      1,
-      'Creating ShaderInputs with modules',
-      allModules.map(m => m.name)
-    )();
+  constructor(modules: {[P in keyof ShaderPropsT]?: ShaderModuleInputs<ShaderPropsT[P]>}) {
+    // Extract modules with dependencies and resolve
+    const allModules = {};
+    _getDependencyGraph({
+      modules: Object.values(modules).filter(module => module.dependencies),
+      level: 0,
+      moduleMap: allModules,
+      moduleDepth: {}
+    });
+    modules = {...modules, ...allModules};
+    log.log(1, 'Creating ShaderInputs with modules', Object.keys(modules))();
 
     // Store the module definitions and create storage for uniform values and binding values, per module
-    this.modules = modules;
+    this.modules = modules as typeof this.modules;
```

Comment:
This seems hacky. Should be properly typed in a shadertools.

Reply:
Are you saying you'd prefer:

```ts
this.modules = modules as Readonly<{
  [P in keyof ShaderPropsT]: ShaderModuleInputs<ShaderPropsT[P]>;
}>;

```